evap/staff/views.py

```diff
@@ -626,40 +652,65 @@ def course_participant_import(request, semester_id, course_id):
     raise_permission_denied_if_archived(course)
 
-    form = UserImportForm(request.POST or None, request.FILES or None)
+    excel_form = UserImportForm(request.POST or None, request.FILES or None)
+    copy_form = CourseParticipantCopyForm(request.POST or None, request.FILES or None)
 
-    if form.is_valid():
+    if request.method == "POST":
         operation = request.POST.get('operation')
-        if operation not in ('test', 'import'):
-            raise SuspiciousOperation("Invalid POST operation")
 
-        # Extract data from form.
-        excel_file = form.cleaned_data['excel_file']
-        import_course = form.cleaned_data['course']
+        if operation not in ('test', 'import', 'copy'):
+            raise SuspiciousOperation("Invalid POST operation")
 
         test_run = operation == 'test'
+        import_run = operation == 'import'
+        copy_run = operation == 'copy'
+
+        if test_run and excel_form.is_valid():
+            excel_file = excel_form.cleaned_data['excel_file']
+            imported_users, warnings, errors = UserImporter.process(request, excel_file, test_run)
+            if not errors:
+                save_import_file(excel_file, str(request.user.id))
+                # messages.success(request, "%d Participants would be added to course %s" % (len(imported_users), course.name))
+                return render(request, "staff_course_participant_import.html", dict(
+                    course=course, excel_form=excel_form, copy_form=copy_form, semester=semester, warnings=warnings, participant_test_passed=True))
+            else:
+                delete_import_file(str(request.user.id))
+                return render(request, "staff_course_participant_import.html", dict(
+                    course=course, excel_form=excel_form, copy_form=copy_form, errors=errors, warnings=warnings, semester=semester, participant_test_passed=False))
+
+        elif import_run:
+            import_filename = get_import_file_name_if_exists(str(request.user.id))
+            if import_filename:
+                imported_users, errors, warnings = UserImporter.process(request, import_filename, test_run)
+                delete_import_file(str(request.user.id))
+                # fallthrough to end
+            else:
+                raise SuspiciousOperation("No test run performed previously.")
+
+        elif copy_run and copy_form.is_valid():
+            import_course = copy_form.cleaned_data['course']
+            if import_course is not None:
+                imported_users = import_course.participants.all()
+                # fallthrough to end
+            else:
+                errors = ["You have to select one course from the drop-down."]
+                return render(request, "staff_course_participant_import.html", dict(
+                    course=course, excel_form=excel_form, copy_form=copy_form, errors=errors, semester=semester, participant_test_passed=False))
```

Comment:
please check if you can use more fallthroughs and have `render(request, "staff_course_participant_import.html", ...)` in fewer, ideally only one place.